omnigan/trainer.py

```diff
@@ -772,64 +808,141 @@ def get_masker_loss(self, multi_domain_batch):  # TODO update docstrings
                             self.logger.losses.generator.task_loss[update_task][
                                 "advent"
                             ][batch_domain] = update_loss.item()
-                elif update_task == "m":
+                elif update_task in {"m", "m2"}:
                     # ? output features classifier
-                    prediction = self.G.decoders[update_task](self.z)
-                    if batch_domain == "s":
 
+                    # If multi-level flooding
+                    if "m2" in self.opts.tasks:
+                        # Get label
+                        if update_task == "m":
+                            prediction = self.G.decoders["m"](
+                                torch.cat((self.z, self.label_1), dim=1)
+                            )
+
+                            # AC-GAN loss
+                            validity, auxiliary = self.D["m2"]["FloodLevel"](
+                                prediction, self.z
+                            )
+
+                            # GAN loss
+                            update_loss = (
+                                self.losses["D"]["default"](validity, True)
+                                * self.opts.train.lambdas.G.m2.gan
+                            )
+                            self.logger.losses.generator.task_loss["m"]["gan"][
+                                batch_domain
+                            ] = update_loss.item()
+
+                            step_loss += update_loss
+
+                            # Auxiliary loss
+                            update_loss = (
+                                self.losses["D"]["multilevel"](
+                                    auxiliary, self.label_1[:, 0, 0, 0].squeeze()
+                                )
+                                * self.opts.train.lambdas.G.m2.aux
+                            )
+
+                            self.logger.losses.generator.task_loss["m"]["aux"][
+                                batch_domain
+                            ] = update_loss.item()
+
+                            step_loss += update_loss
+
+                        elif update_task == "m2":
```

Comment:
If I read this right, the only things that change between the `if` and `elif` are `self.label_1[:, 0, 0, 0].squeeze()` vs `self.label_2[:, 0, 0, 0].squeeze()` and `self.logger.losses.generator.task_loss`. I bet you can refactor this whole block in a much shorter way by having variables and a common code dependent on those. Cleaner, shorter, less error-prone (don't need to change 2 pieces of code if you change the logic) more versatile (what about more flood levels?)